### PR TITLE
[Merged by Bors] - eth2_hashing: make `cpufeatures` dep optional

### PR DESCRIPTION
**Pull request by sept-en**
## Issue Addressed

#3308 

## Proposed Changes

* add `cpufeatures` feature.
* make `cpufeature` default feature to preserve the compatibility;
* hide all `cpufeature`-related code with `cpufeatures` feature.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Awesome thanks!

Do you mind amending your commit and force pushing it to re-trigger Github's CI? There's a glitch where Github won't re-run jobs after the target branch is changed (I've changed your PR to target `unstable`).

**Comment by michaelsproul** [comment]
You might also have to differentiate the feature from the dep name, as we're still maintaining Rust 1.58 compatibility. I'd be happy with a feature called `detect-cpufeatures` or similar.

**Comment by sept-en** [comment]
@michaelsproul sure changed this. Please check

**Review by michaelsproul — APPROVED** [review]
Looks great, thank you! Will merge once CI has passed

File: crypto/eth2_hashing/src/lib.rs [hunk 1]
```diff
@@ -127,16 +127,15 @@ pub enum DynamicImpl {
...
+#[cfg(all(feature = "detect-cpufeatures", target_arch = "x86_64"))]
 cpufeatures::new!(x86_sha_extensions, "sha", "sse2", "ssse3", "sse4.1");
 
 #[inline(always)]
 pub fn have_sha_extensions() -> bool {
-    #[cfg(target_arch = "x86_64")]
+    #[cfg(all(feature = "detect-cpufeatures", target_arch = "x86_64"))]
     return x86_sha_extensions::get();
 
-    #[cfg(not(target_arch = "x86_64"))]
```
Comment (michaelsproul) [inline crypto/eth2_hashing/src/lib.rs:138]: We need to keep a `cfg` here to prevent the compiler from warning about unreachable code

Comment (michaelsproul) [reply crypto/eth2_hashing/src/lib.rs:138]: we want the negation of the other condition: `cfg(not(all(feature = "detect-cpufeatures", target_arch = "x86_64")))`

Comment (sept-en) [reply crypto/eth2_hashing/src/lib.rs:138]: Yeah, that makes sense! Fixed in be49f5b

**Comment by michaelsproul** [comment]
The cargo audit failures are expected, and will be fixed by another PR in the pipeline. We can merge this as soon as the tests pass.

**Comment by michaelsproul** [comment]
CI being flaky. C'mon!

bors retry

**Pull request closed without merging**